### ISSUE TITLE
Need some advice on my node code for using the repo.

### ISSUE THREAD

**Issue by zydjohnHotmail** [iamnotstatic/multichain-crypto-wallet#23]
#### Steps to reproduce the behavior

1. [First step]
2. [Second step]
3. [and so on...]

#### Environment
NPM: 9.1.3
Node: 9.2.0
npm multichain-crypto-wallet: 0.2.7

Hello:
I want to create one wallet with address for receiving RUFF (Ruff-Chain) tokens on HECO network, and after checking account balance and less the gas fees, send all the RUFF tokens to another address via ERC20 network.
I can send some Ruff tokens from one exchange via web browser, so this part will work, as I did some test 2 months ago, and it worked.
So, the following is my code to do the job for check account balance:

const data = await multichainWallet.getBalance({
    address: '0x2455eC6700092991Ce0782365A89d5Cd89c8Fa22',
    network: 'ethereum',
    rpcUrl: 'https://http-mainnet.hecochain.com',
  });

Then the following code to send 100 Ruff tokens to another address via ERC20 network.
const transfer = await multichainWallet.transfer({
  recipientAddress: '0x2455eC6700092991Ce0782365A89d5Cd89c8Fa22',
  amount: 100,
  network: 'ethereum',
  rpcUrl: 'https://1rpc.io/eth', => Ethereum Mainnet
  privateKey: 'private key',
  tokenAddress: '0xf278c1ca969095ffddded020290cf8b5c424ace2', => Ruff token contract address
});

Please advice if my code is OK or not.
By the way, how can I find and pay the gas fees?
One issue is: the Ruff token is dirty cheap token, it is difficult to know how much gas fee I should pay.
Is possible that I use the same Wallet address, which can also receive some USDT via HECO network, and pay the gas fees using USDT balance?
Or I just guess the gas fees, for example, if I have total 120 Ruff tokens, and I am sure that the gas fees is less than 20 Ruff tokens, then sending 100 Ruff tokens will make sure the transaction will work as expected?
Please advice as I am going to do this for real soon.
By the way, can you publish a new version NPM for this repo?
Thanks,

**Comment by zydjohnHotmail** [comment]
Hello:
I got the following gas fees from etherscan:
{
	"status": "1",
	"message": "OK",
	"result": {
		"LastBlock": "16131929",
		"SafeGasPrice": "14",
		"ProposeGasPrice": "14",
		"FastGasPrice": "15",
		"suggestBaseFee": "13.367311122",
		"gasUsedRatio": "0.643610966666667,0.9677928,0.341929233333333,0.3364585,0.543919"
	}
}

Please show me how to use it in sending USDT or RUFF tokens using this npm package.
Thanks,

**Comment by iamnotstatic** [comment]
Hi @zydjohnHotmail , You can use the `SafeGasPrice,` `ProposeGasPrice` or `FastGasPrice` it depends on fast you want the transaction to get mined and how much you're willing to pay for the gas fee. Also noted you need ETH to pay for gas fee.

**Comment by zydjohnHotmail** [comment]
Hello:
Thank you for your reply, I studied the repo, and I see the following JavaScript in
https://cdn.jsdelivr.net/npm/multichain-crypto-wallet@0.2.7/src/common/utils/types.ts

export interface TransferPayload {
  recipientAddress: string;
  amount: number;
  network: string;
  rpcUrl?: string;
  privateKey: string;
  gasPrice?: string;
  tokenAddress?: string;
  nonce?: number;
  data?: string;
  gasLimit?: number;
  fee?: number; // defaults to 10000
  subtractFee?: boolean; // defaults to false
}
The fee, defaults to 1000, is it very high? I mean if I don’t provide the fee parameter, the library will use 10000 as fee?
From the Gas fee Json format data I got from EtherScan, can I write the following code to transfer 100 RUFF tokens?

const transfer = await multichainWallet.transfer({
recipientAddress: '0x2455eC6700092991Ce0782365A89d5Cd89c8Fa22',
amount: 100,
network: 'ethereum',
rpcUrl: 'https://1rpc.io/eth', => Ethereum Mainnet
privateKey: 'private key',
tokenAddress: '0xf278c1ca969095ffddded020290cf8b5c424ace2', => Ruff token contract address
fee: 15,
}); 
Let me know if my code is correct according to the gas fees I recently got.
By the way, you said I need ETH to pay the gas fee, then you mean I have to deposit some ETH in the same wallet address before I can send some RUFF or USDT tokens?
Finally, I want to start with HECO testnet, and I see the testnet RPC is:
https://http-testnet.hecochain.com/
I can get some free HT faucet from its faucet URL:
https://scan-testnet.hecochain.com/faucet

But how can I generate a wallet address only used for HECO test net?
Like this:
const address = multichainWallet.getAddressFromPrivateKey({
  privateKey:
'bXXgTj2cgXMFAGpLHkF5GhnoNeUpmcJDsxXDhXQhQhL2BDpJumdwMGeC5Cs66stsN3GfkMH8oyHu24dnojKbtfp',
  network: 'hecochain-test',
});


In short, how can I provide the network parameter, so the generated address will be on HECO testnet?
Please advice,
Thanks,

**Comment by iamnotstatic** [comment]
Nope. the `fee` is for bitcoin for Ethereum, you use the `gasPrice ` param. Yeah, the address you want to send from needs to have the chain native token to pay for the gas fee.

Yeah,  Generated address on Ethereum will also work with the HECO chain all you have to do is use the chain rpc url

**Comment by iamnotstatic** [comment]
> 0xf278c1ca969095ffddded020290cf8b5c424ace2

Yeah, the code on the Ruff token transfer is correct. I can't tell because the gas fee is not static,  but you can always check the wallet's balance before transferring to be sure you have enough funds for the gas price.

**Comment by iamnotstatic** [comment]
> Hello:
> 
> I succeeded in receiving some HECO testnet faucets, the following is my code and result output: D:\nodejs\MultiChainCrypto>type ETH_Wallet_Balance.js const multichainWallet = require('multichain-crypto-wallet');
> 
> const accountBalance = async () => { try { const data = await multichainWallet.getBalance({ address: '0xbC4EcBe1b0d543126c441De5364F122B78d90972', network: 'ethereum', rpcUrl: 'https://http-testnet.hecochain.com/', }); const json_balance = JSON.stringify(data); console.log(json_balance); } catch (err) { console.log(err); } };
> 
> accountBalance(); D:\nodejs\MultiChainCrypto>node ETH_Wallet_Balance.js bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?) {"balance":0.5}
> 
> I can see from the faucet web page that the faucet sent me 0.5HT (Huobi Token) to my wallet address. However, the code I used can only see the amount of the token, but how can I show the token name, in this case, it is HT. Please advise, Thanks,

This is a native token you can prefix the coin name based on the chain, But for tokens, you can call the `name`, and `symbol` methods in the contract.

**Comment by iamnotstatic** [comment]
> Hello:
> As I can't find more details about this, could you please share some code on how to use: name and symbol in the node script?
> For gas fees, after I study, if the gasPrice = 15 for each transaction (if the 15 means 15 gwei), for 1USDT worth of ETH, it can be used for 54,000 transactions. So if I deposit 10USDT worth of ETH, then it can be used for quite a few month. Am I right?
> Thanks,

Possibly but you need to be aware that gas fee changes based on  network congestion. 

You use the custom smart contract call method. 

```
const data = await multichainWallet.smartContractCall({
  rpcUrl: 'https://rpc.ankr.com/eth',
  network: 'ethereum',
  contractAddress: '0xdAC17F958D2ee523a2206206994597C13D831ec7',
  method: 'name',
  methodType: 'read',
  params: [],
  contractAbi: [
   contractAbi: [
     {
    "constant": true,
    "inputs": [],
    "name": "name",
    "outputs": [{ "name": "", "type": "string" }],
    "payable": false,
    "stateMutability": "view",
    "type": "function"
  },
  ],
});
```

**Comment by zydjohnHotmail** [comment]
Hello:
Thanks for your quick reply, I will spend some time to study this, this seems not very easy to use.
However, I have a question to code for showing the currency name with its contract address. Let's say I have one wallet address, and I want to deposit about 10 different currency, each with its own contract address.  If I do NOT want to write 10 different scripts to query currency balance with each script for one currency.
Let's assume all 10 tokens are on the same chain.
Can I use something like dictionary in C#, use one Currency Name as key, like: HT, or USDT or BTC; then use the contract address as value.
I mean for example the above code:
I assume the currency is 'HT' as the key, and value is the contractAddress: '0xdAC17F958D2ee523a2206206994597C13D831ec7',
So I want to make a loop for all 10 differenet currencies with different contract addres, show different account balance?
Please advise,
Thanks,

**Comment by zydjohnHotmail** [comment]
Hello:
Thanks for your code, but I found some issues to use your code to show currency names.
Your code did not have wallet address, which is necessary to get account balance, so I added it. But I found that your code used twice: contractAbi: 

I use Visual Studio 2022 to create a Node JS application, and paste the code, VS 2022 complains that there should be a ‘,’ before the second contractAbi: 
Therefore, I deleted the second contractAbi: 

The following is my code now:
const multichainWallet = require('multichain-crypto-wallet');

const accountHTBalance = async () => {
    try {
        const data = await multichainWallet.getBalance
            ({
            address: '0xbC4EcBe1b0d543126c441De5364F122B78d90972',
            network: 'ethereum',
            rpcUrl: 'https://http-testnet.hecochain.com/',
            contractAddress: '0x6f259637dcd74c767781e37bc6133cd6a68aa161', => HT contract address in EtherScan
            method: 'name',
            methodType: 'read',
            params: [],
            contractAbi: [{
            "constant": true,
            "inputs": [],
            "name": "name",
            "outputs": [{ "name": "", "type": "string" }],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
            }]});
        const json_balance = JSON.stringify(data);
        console.log(json_balance);
    }
    catch (err) {
        console.log(err);
    }
};

accountHTBalance();

When I run this, I got the following output:
D:\nodejs\MultiChainCrypto>node ETH_Wallet_Check_HT_Balance.js
bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)
{"balance":0.5}


I think maybe this is due to the fact that in Test net, such contract address does NOT exist?
Please advice if I should use contractAbi: twice in the code, but if I use it, then I have issue in VS 2022, which complains syntax issue.
Thanks,

**Comment by zydjohnHotmail** [comment]
Hello:

I found more issues when trying to send some HT faucet tokens via HECO test net.
The following is my code: (I have received 0.5HT from HECO test net faucet), and after checking account balance, I would like to send 0.1HT from my account balance. I created a new wallet address to receive the HT tokens.
const multichainWallet = require('multichain-crypto-wallet');

const transferHT = async () => 
{
    try 
	{
	const transfer = await multichainWallet.transfer({
	recipientAddress: '0x920C2Cd0eEd6c393A3517f080799F02AAfDBC084', => New address
	amount: 0.1,
	network: 'ethereum',
	rpcUrl: 'https://http-testnet.hecochain.com/',
	privateKey: 'My_Private_Key',
	tokenAddress: '0x6f259637dcd74c767781e37bc6133cd6a68aa161', => HT contract address
	});
	const json_transfer = JSON.stringify(transfer);
	console.log(json_transfer);
    }
    catch (err) 
	{
        console.log(err);
    }
};

transferHT();

When I run my code, I got the following errors:
D:\nodejs\MultiChainCrypto>node Send_HT_To_Address.js
bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)
Error: call revert exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="decimals()", data="0x", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXCEPTION, version=abi/5.7.0)
    at Logger.makeError (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\logger\lib\index.js:238:21)
    at Logger.throwError (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\logger\lib\index.js:247:20)
    at Interface.decodeFunctionResult (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\abi\lib\interface.js:388:23)
    at Contract.<anonymous> (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\contracts\lib\index.js:395:56)
    at step (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\contracts\lib\index.js:48:23)
    at Object.next (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\contracts\lib\index.js:29:53)
    at fulfilled (D:\nodejs\MultiChainCrypto\node_modules\@ethersproject\contracts\lib\index.js:20:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  reason: null,
  code: 'CALL_EXCEPTION',
  method: 'decimals()',
  data: '0x',
  errorArgs: null,
  errorName: null,
  errorSignature: null,
  address: '0x6f259637dcd74c767781e37bc6133cd6a68aa161',
  args: [],
  transaction: {
    data: '0x313ce567',
    to: '0x6f259637dcD74C767781E37Bc6133cd6A68aa161',
    from: '0x920C2Cd0eEd6c393A3517f080799F02AAfDBC084'
  }
}

D:\nodejs\MultiChainCrypto>

**Comment by iamnotstatic** [comment]
> 0x6f259637dcd74c767781e37bc6133cd6a68aa161

Looking at your code, what you’re passing as `tokenAddress` is not a smart contract address that’s a EOA, that’s why it can’t make get decimal call on the contract

**Comment by zydjohnHotmail** [comment]
Hello:
Fortunately, I asked Huobi Wallet technical support a few days ago about the contract address, the following was their reply:
+++++++++++++++++++++++++++++++++++++++++++++++++++
T2-JoJo (Huobi Wallet)
Dec 7, 2022, 17:52 GMT+8
Dear Customer,
Thank you for contacting Huobi Customer Service Center.
This e-mail is in response to your below questions:
Hello, if you are going to Huobi deposit HT. The contract address of the ERC20 chain is 0x6f259637dcd74c767781e37bc6133cd6a68aa161, and the HT of the HECO chain does not have a contract address, please be informed, thank you.
+++++++++++++++++++++++++++++++++++++++++++++++++++++
From their reply, HT on their HECO chain doesn't have a contract address, as I can't find necessary information from HECO explorer, the contract address is the ERC20 chain. In this case and many other HECO tokens, if they don't have contract address, then what I should do?
By the way, for showing currency name or symbol, like 'HT' on account balance, my code did not work.
Any more suggestion?
Please advise,

**Comment by zydjohnHotmail** [comment]
Hello:
Just as I said before: I want to be able to receive tokens from HECO chain, and after paying the gas fees, send the received token to another address via ERC20 network.
Before I go to main net, I want to use test net to see if I can receive tokens from HECO chain, and then send the received tokens to another address via ERC20 network.
In my testing case, I can receive test net HT token via faucet sent by HECO test net, so it works for the first part; now the second part is: I want to send the received HT tokens to another address via ERC20 (Ethereum) test net.
Does my logic works in the test net?
I think if my testing works for test net, then on main net, it should work too.
Please advice,

**Comment by iamnotstatic** [comment]
Okay, I think first of all you need to understand how Heco and Ethereum chain works; you can't directly send funds received from HECO to Ethereum they run on different environments; So, for you to be able to do that, you need a Bridge which I sent to you a couple of days ago. 

But what this package can help you do is easily interact with these chains seamlessly.

**Comment by iamnotstatic** [comment]
> Hello: OK, if I can't do this via this package. Then let me know if I can send tokens received from solana network and send to ERC20 network using this package? Thanks,

Hi @zydjohnHotmail , Just to make clarity what this package does is give you the abity to interact with different blockchains and smart contracts. 

What you're trying to achieve can't be directly done with the package, so you will need a bridge to be able to bridge funds from one blockchain to another, you can read more about bridges here https://101blockchains.com/blockchain-bridges/

**Comment by zydjohnHotmail** [comment]
Hello:
I studied about bridge, I think for bridge between Solana and EVM, wormhole can do  the job.
My question is: how can I use this package to let wormhole to bridge funds, I see wormhole also has some JavaScript SDK.  But it seems I have to setup a wallet and do a lot of job by hand.
If I can use this package to generate wallet, then can I use the package to let wormhole to bridge fund without a lot of job by click on web page.  For setup or configuration, it is OK.  But I don't like to do a lot of mouse click each time for bridge funds, as if the first time bridge fund works, there could be a lot of such things coming.
Please advise,
Thanks

**Comment by iamnotstatic** [comment]
> Hello: I studied about bridge, I think for bridge between Solana and EVM, wormhole can do the job. My question is: how can I use this package to let wormhole to bridge funds, I see wormhole also has some JavaScript SDK. But it seems I have to setup a wallet and do a lot of job by hand. If I can use this package to generate wallet, then can I use the package to let wormhole to bridge fund without a lot of job by click on web page. For setup or configuration, it is OK. But I don't like to do a lot of mouse click each time for bridge funds, as if the first time bridge fund works, there could be a lot of such things coming. Please advise, Thanks

Yeah, You can interact with the wormhole smart contract using the package `smartContractCall.`  You can learn more about interacting with smart contract.

**Comment by zydjohnHotmail** [comment]
Hi,
Do you have some code example or some web site on this topic?
By the way, I used your code trying to show currency name, like 'HT' did not work.
Can you find what went wrong?
Thanks,

**Comment by zydjohnHotmail** [comment]
The following is my code now:
const multichainWallet = require('multichain-crypto-wallet');

const accountHTBalance = async () => {
try {
const data = await multichainWallet.getBalance
({
address: '0xbC4EcBe1b0d543126c441De5364F122B78d90972',
network: 'ethereum',
rpcUrl: 'https://http-testnet.hecochain.com/',
contractAddress: '0x6f259637dcd74c767781e37bc6133cd6a68aa161', => HT contract address in EtherScan
method: 'name',
methodType: 'read',
params: [],
contractAbi: [{
"constant": true,
"inputs": [],
"name": "name",
"outputs": [{ "name": "", "type": "string" }],
"payable": false,
"stateMutability": "view",
"type": "function"
}]});
const json_balance = JSON.stringify(data);
console.log(json_balance);
}
catch (err) {
console.log(err);
}
};

accountHTBalance();

When I run this, I got the following output:
D:\nodejs\MultiChainCrypto>node ETH_Wallet_Check_HT_Balance.js
bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)
{"balance":0.5}

**Comment by zydjohnHotmail** [comment]
By the way, I found at least 2 command line wallets which work for Windows 10.
Then I think maybe I can use the package with the command line wallets and send/transfer fund via Wormhole or another bridge: Celer cBridge.
Please advise on the data flow for this, my target is to use API/SDK or CLI, not always use mouse clicks on web page.
Thanks,

**Comment by iamnotstatic** [comment]
> The following is my code now: const multichainWallet = require('multichain-crypto-wallet');
> 
> const accountHTBalance = async () => { try { const data = await multichainWallet.getBalance ({ address: '0xbC4EcBe1b0d543126c441De5364F122B78d90972', network: 'ethereum', rpcUrl: 'https://http-testnet.hecochain.com/', contractAddress: '0x6f259637dcd74c767781e37bc6133cd6a68aa161', => HT contract address in EtherScan method: 'name', methodType: 'read', params: [], contractAbi: [{ "constant": true, "inputs": [], "name": "name", "outputs": [{ "name": "", "type": "string" }], "payable": false, "stateMutability": "view", "type": "function" }]}); const json_balance = JSON.stringify(data); console.log(json_balance); } catch (err) { console.log(err); } };
> 
> accountHTBalance();
> 
> When I run this, I got the following output: D:\nodejs\MultiChainCrypto>node ETH_Wallet_Check_HT_Balance.js bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?) {"balance":0.5}

This is not what I showed you, you're interacting with `getBalance` and expecting token name,  You're to interact with the https://github.com/iamnotstatic/multichain-crypto-wallet#smart-contract-call

**Comment by iamnotstatic** [comment]
> By the way, I found at least 2 command line wallets which work for Windows 10. Then I think maybe I can use the package with the command line wallets and send/transfer fund via Wormhole or another bridge: Celer cBridge. Please advise on the data flow for this, my target is to use API/SDK or CLI, not always use mouse clicks on web page. Thanks,

You can look for how to interact with their smart contract.
